### DISCUSSION TITLE
nested value disappear when remount NestedFieldArray

### DISCUSSION THREAD

**Discussion by TseHang** [react-hook-form/react-hook-form#4417] [Q&A] [answered]
**Describe the bug**
I have a nested array form, and some components need to be hidden when `show` flag is false.
The form will remove component when show is false, and mount component when flag is true.
It is perfect at the first level field array. However, in nested-field-array, we lose the default values after we re mount it.

I am not sure it is a bug or I misunderstood the correct way. 😭
Please help me!

**To Reproduce**

1. Go to [code sandbox link](https://codesandbox.io/s/react-hook-form-nestedarray-empty-issue-fpolt?file=/src/Item.js)
2. Click "show" checkbox in first level
3. See our nested array components.
3. Click test1's "show" checkbox.
4. See test1's children(nested field array)
5. Click test1's "show" checkbox.
6. Hidden test1's children(nested field array)
7. Click test1's "show" checkbox.
8. **LOL... We can not see the test1's children(nested field array)**

**Codesandbox link (Required)**
Include a codesandbox will help us to investigate the issue quicker.
https://codesandbox.io/s/react-hook-form-nestedarray-empty-issue-fpolt?file=/src/Item.js

**Expected behavior**
1. Go to [code sandbox link](https://codesandbox.io/s/react-hook-form-nestedarray-empty-issue-fpolt?file=/src/Item.js)
2. Click "show" checkbox in first level
3. See our nested array components.
3. Click test1's "show" checkbox.
4. See test1's children(nested field array)
5. Click test1's "show" checkbox.
6. Hidden test1's children(nested field array)
7. Click test1's "show" checkbox.
8. **See the test1's children(nested field array)**

**Screenshots**
![螢幕錄製 2021-03-13 下午3 06 57](https://user-images.githubusercontent.com/14778824/111022698-5a57a000-840f-11eb-953b-7c6d24e658f9.gif)

**Desktop (please complete the following information):**

- OS: macOS Big Sur 11.0.1
- Browser: Safari 14.0.1
- React Hook Form Version: 16.15.4

**Additional context**
I am not sure it is a bug or I misunderstood the correct way. 😭
IMO, I should get defaultValue when mount a fieldArray component. And it works on the version 6.14.2
https://codesandbox.io/s/react-hook-form-nestedarray-empty-issue16142-wrznk?file=/src/Item.js

**Comment by bluebill1049** [comment]
for v6, you can set config `shouldUnregister: false`. this will be the default behavior in v7.

**Comment by TseHang** [comment]
If it is possible, I have another questions about this example: https://codesandbox.io/s/react-hook-form-nestedarray-empty-issue16154-fpolt

1. Although I watch in the `Item.js` component, looks like it triggers the whole field array... Why? However, it works expectedly  when using `useWatch`.
2. `renderCount` always add 2 when I click `show` flag, however, only get one log in console... curiously 😭

**Comment by TseHang** [comment]
> for v6, you can set config `shouldUnregister: false`. this will be the default behavior in v7.

Thanks. I got the answer from the discussion!

**Comment by bluebill1049** [comment]
watch is always at root, useWatch at component level.

**Comment by TseHang** [comment]
I have a little strange about  why my link can work on `6.14.2`? but I can not work on `6.15.4`?

6.15.4: https://codesandbox.io/s/react-hook-form-nestedarray-empty-issue16154-fpolt

6.14.2: https://codesandbox.io/s/react-hook-form-nestedarray-empty-issue16142-wrznk

---
In my example `6.15.4`,
At root level's `show flag`: It works with defaultValue after unmount.
At children nested level's `show flag`: It can not get defaultValue after unmount.
What is the reason?

![螢幕錄製 2021-03-13 下午4 24 24](https://user-images.githubusercontent.com/14778824/111024261-b3780180-8418-11eb-9bab-34769a0b8d78.gif)


**I am not sure, but maybe my real problem is why I can not get defaultValues in nested-field-array after re-mount?**

p.s. In `6.14.2`, I don't set `shouldUnregister: false`, and I can get the defaultValues in root & nested children level.

**Comment by TseHang** [comment]
> watch is always at root, useWatch at component level.

Thanks!!!!

> 2. `renderCount` always add 2 when I click `show` flag, however, only get one log in console... curiously 😭

Just don't know why, could somebody tell me why ? 😭

**Comment by bluebill1049** [comment]
for unmounted inputs to be retained with value, use `shouldUnreigster: false` otherwise consider upgrading to v7 as it's in RC now.

**Comment by TseHang** [comment, marked as answer]
Curiously, how if I want to set value to a re-registered component ?

I use setValue in this way is not worked 😢 (just see `Item.js`)
https://codesandbox.io/s/react-hook-form-nestedarray-empty-issue16154-setvalue-btsqs

**Reply by bluebill1049** [reply]
Just did a quick conversation to v7: https://codesandbox.io/s/react-hook-form-nestedarray-empty-issue16154-setvalue-forked-1l6t5?file=/src/App.js 

let me know if this is better.

**Reply by bluebill1049** [reply]
<img width="526" alt="Screen Shot 2021-03-13 at 9 20 08 pm" src="https://user-images.githubusercontent.com/10513364/111027043-e4206100-8441-11eb-85be-8b20893555e6.png">
also you checkbox name is not in field array shape.

input name needs to be unique.

**Reply by TseHang** [reply]
@bluebill1049 Thanks your quick conservation.

I fix Controller and migrate into v7 here:
https://codesandbox.io/s/react-hook-form-nestedarray-empty-issuev7-setvalue-f0uhl

The scenario I want is:
1. Don't submit unmounted value in UI
2. Get defaultValue when I mount the component

I try to complete 2. by useEffect here
see `Items.js`
see `NestedItems.js`
<img width="560" alt="截圖 2021-03-14 上午10 36 45" src="https://user-images.githubusercontent.com/14778824/111055482-36e53180-84b1-11eb-8cbb-972dd0a7036e.png">

If the way is not ideal please let me know.
Thanks very much with your kindly response.

**Reply by bluebill1049** [reply]
> 1) Don't submit unmounted value in UI

what do you mean by unmounted value, are those values a part of the `defaultValues` or value gets appended but not in the UI?

**Reply by TseHang** [reply]
> Don't submit unmounted value in UI

I think **"only submit the values on UI"** is more precise.

So that I think defaultValues should not include those values not exist in UI at first time , and append values when component mounted ?

**Reply by bluebill1049** [reply]
we have already agreed to include `defaultValues` as part of the submission data during v7 RFC (with most of the users requested for such behavior), so we are not going to change that.

**Reply by bluebill1049** [reply]
you can readmore here: https://github.com/react-hook-form/react-hook-form/discussions/3714

**Reply by TseHang** [reply]
Yes, I have already know the decision, and thanks your link. 👍

My question here https://github.com/react-hook-form/react-hook-form/discussions/4417#discussioncomment-479526 is if I only want to submit the values on UI, the way is I should `setValue` hidden form components in `useEffect` and unregister it when unmount. Is it a recommend way or not?

want to know if it is a good way when using react-hook-form, thanks!

**Reply by bluebill1049** [reply]
sounds legit to me, I would `register` them explicitly as well.

```
React.useEffect(() => {
  register('test');
  setValue('test', 'test');
})
```

**Reply by TseHang** [reply]
👍 Great! Thanks
